cf-check/backup.c

```diff
@@ -0,0 +1,198 @@
+#include <platform.h>
+#include <backup.h>
+
+#if defined (__MINGW32__)
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup not supported on Windows\n");
+    return 1;
+}
+
+#elif ! defined (LMDB)
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup only implemented for LMDB.\n");
+    return 1;
+}
+
+#else
+
+#include <stdio.h>
+
+bool copy_file(const char *src, const char *dst)
+{
+    assert(src != NULL);
+    assert(dst != NULL);
+
+    printf("Copying: '%s' -> '%s'\n", src, dst);
+    FILE *in = fopen(src, "r");
+    if (in == NULL)
+    {
+        return false;
+    }
+    FILE *out = fopen(dst, "w");
+    if (out == NULL)
+    {
+        fclose(in);
+        return false;
+    }
+    size_t read = 0;
+    size_t written = 0;
+    bool ret = true;
+    do
+    {
+        const size_t bufsize = 1024;
+        char buf[1024] = { 0 };
+
+        read = fread(buf, sizeof(char), bufsize, in);
+
+        written = fwrite(buf, sizeof(char), read, out);
+        while ((!ferror(out)) && (written < read))
+        {
+            written += fwrite(buf + written, sizeof(char), read - written, out);
+        }
+
+        if (read != written || ferror(in) || ferror(out))
+        {
+            break;
```

Comment:
again, just curious if you can use errno here. It seems that fwrite/fread/ferror won't set errno directly but underlying read/write could.

Reply:
No, I think it's better to follow what manpage says:

```
The function fread() does not distinguish between end-of-file and error; callers must use feof(3)
and ferror(3) to determine which occurred.  The function fwrite() returns a value less than nitems
only if a write error has occurred.
```